### ISSUE TITLE
Docker container image

### ISSUE THREAD

**Issue by disrupted** [piejanssens/premiumizer#88]
Is there a way to use this with Docker? now Docker doesn't need an introduction but just let me say it streamlines the whole setup process and lets all services run in a sandbox environment which is super cool

**Comment by disrupted** [comment]
I tried packaging up as a Docker image myself but I am missing something. I followed this tutorial and created a `Dockerfile` which looks like this
```dockerfile
FROM python:2.7
RUN git clone git://github.com/piejanssens/premiumizer.git
RUN pip install -r premiumizer/requirements.txt
WORKDIR /premiumizer
EXPOSE 5000
VOLUME /premiumizer /config
CMD ["python", "premiumizer.py"]
```
I could create the image successfully then by running `docker build -t disrupted/premiumizer .`. When I run the container using `docker run --name premiumizer -p 5000:5000 disrupted/premiumizer` I get the usual output but I can't connect to the webserver (even though I bridged port 5000). perhaps someone can help, shouldn't be too hard.

```log
11-10 22:45:24: INFO : *************************************************************************************
11-10 22:45:24: INFO : ----------------Premiumizer.log file has been deleted as a precaution----------------
11-10 22:45:24: INFO : *************************************************************************************
11-10 22:45:24: INFO : *************************************************************************************
11-10 22:45:24: INFO : ---------------------------Premiumizer has been updated!!----------------------------
11-10 22:45:24: INFO : *************************************************************************************
11-10 22:45:24: INFO : Running at /premiumizer
11-10 22:45:24: INFO : Starting server on 127.0.0.1:5000
11-10 22:45:34: ERROR : premiumize.me login error: {"status":"error","message":"Not logged in. Please log in first."}
```

**Comment by bjsjr** [comment]
How are you trying to access the webserver?  You can't use 127.0.0.1:5000 from outside of the container because you'll be trying to connect to the host's localhost, not the container's.

I'm not very familiar with docker but hopefully that will help you find an answer.

**Comment by Killerpretzel** [comment]
Hey @disrupted,

I had a very similar problem. The browser says 'empty response' or something like that?

Try editing the 'settings.cfg' file of Premiumizer. You need to change the 'bind_ip' option to the actual IP of the network-interface within the Docker-container.

**Comment by H4R0** [comment]
@disrupted 
You have to replace the listening ip from 127.0.0.1 to 0.0.0.0 (bind to all interfaces)
You can do this after the git clone, should be something like:
RUN sed -i "s/127.0.0.1/0.0.0.0/g" premiumizer/settings.cfg.tpl

**Comment by disrupted** [comment]
@H4R0 @Killerpretzel that did it! thanks

**Comment by H4R0** [comment]
@disrupted I'm looking forward to it.
I just want to consider some things i implemented in my premiumizer Docker image.
- Use alpine (FROM pythoin:2.7-alpine) which will result in a way smaller image 25MB Image size vs 273MB. You may have to install git because its not included in the alpine image.
- As you want to share the files downloaded by premiumizer with other containers file permission can mess up quickly. Premiumizer will run as root by default and creates files as read only for other users. You probably can use the -u option to change the uid on run. Other Containers for example linuxserver.io implemented this using PUID/PGID env vars. Another way would be to change the umask. Just test if the downloaded files are read/writable by other services, if theay are its fine.

My setup is running since 4 months, i use premiumizer as blackhole download client for Sonarr and Radarr in combination with Jackett. I only had to apply a simple patch to get all torrent trackers working.

**Comment by Killerpretzel** [comment]
Hey @H4R0

you run a setup similar to mine, maybe you can give me a quick hint (or someone else, of course).

Sonarr puts the files into a specific folder, Premiumizer picks them up and they get downloaded into the cloud storage. So far, so good. 

However, quite often the corresponding category doesn't get assigned and the files just sit there with "Waiting on category". 

Do you use the category-system, or did you just activate the option "Download everything and assign the category default"?

I could do that, of course, but sometimes I want to manually download stuff with Premiumise.me. Additionally, I plan to install other tools similar to Sonarr later on, that would also use this folder structure. I'm concerned clumping everything into the same input- and output-folder will lead to additional problems down the line.

How should I handle this? Am I missing something?

**Comment by neox387** [comment]
If the directory name where Sonarr puts the file matches with the one in your category settings it should apply that category.

You say 'However, quite often' so from the same folder sometimes it works & sometimes it doesn't ? That would be weird.

Premiumizer uses:
os.path.basename(os.path.normpath(os.path.dirname(watchdir_file)))

To get the directory name to match with the category name, the folders for watchdir/blackhole should be automatically created if they don't exist when starting premiumizer.

Only thing I can think of is some non ascii characters in the path or something..

**Comment by Killerpretzel** [comment]
Hey @neox387

> You say 'However, quite often' so from the same folder sometimes it works & sometimes it doesn't ? That would be weird.

That's exactly what happens.
Sometimes the categories are added without a problem, sometimes not. I found no pattern as of yet.

I took a look into the log and it actually says "added with category tv", however, the download has no category assinged in the list.

> Only thing I can think of is some non ascii characters in the path or something..

As far as I can tell, the particular download task I looked at, has no weird characters in the name, not even spaces, just periods. I observed several other tasks that also looked alright. After you asign the catergory manually, the download starts.

I actually deleted the database-file and started over, because I thought I messed something up, but this problem kept occurring soon after I set everything back up.

Would the debug mode give some more insight into this?

**Comment by H4R0** [comment]
@Killerpretzel i had the exact same problem with the newest version of premiumizer after i upgraded. I'm currently running the older version again because i couldn't find a solution for it. Even using the new option "assign default category" or how it is named had no effect, premiumizer just wouldnt download them automatically. One thing i realized in the History is that after premiumizer picks up the downloads from the blackhole directory the Category is correctly displayed in the History log. Premiumizer then uploads the Task to premiumize and waits for it to finish. On the next History log entrie the categoriy is gone and the file name also changed. I think it has todo with spaces somehow. I remember that spaces were first displayed correctly but after premiumize download they had been replaced by "+" signs. Maybe premumize changes the filename / encoding ? I will reproduce the behavior and copy the Log.

**Comment by H4R0** [comment]
Weird i can not reproduce the behavior. I tried 7 different downloads and everything works. But i had the exact same problem as you mentioned last time.

**Comment by neox387** [comment]
@H4R0 there was or maybe still is a problem with the naming, this is because some downloads that premiuimize.me doesn't have yet don't get a name until it start downloading, so premiumize.me response with name = Loading name

and then premiumizer changes the name once premiumize.me knows the name, something might go wrong there, should change this to compare hashes.

@Killerpretzel Maybe you can confirm this; the sometimes when it doesn't work it's probably 'loading name' on the premiuimze.me website, check your log file the added: name of torrent might be different then the name when it's ready to download with no category ?

**Comment by Killerpretzel** [comment]
> Maybe you can confirm this; the sometimes when it doesn't work it's probably 'loading name' on the premiuimze.me website, check your log file the added: name of torrent might be different then the name when it's ready to download with no category ?

@neox387 I have enabled the debug-log and poked at it some more.

Sadly, I can't really confirm anything. I have seen cases where the name stays exactly the same, sometimes Premiumize.me displays an entirely different name, sometimes the name seems to be truncated. In all cases however, at times the categories work and sometimes they don't. I still haven't found a pattern.

I didn't really look at the "loading name" thing, I know what you mean, but I haven't found any hints in the logs about it.

> should change this to compare hashes.

That sounds awesome, but also like a lot of work.

> i had the exact same problem with the newest version of premiumizer after i upgraded. I'm currently running the older version again because i couldn't find a solution for it.

@H4R0 I'm glad someone knows what I'm talking about.

The older version might be worth a shot, but I am a huge Docker newb and it took me quite some time to get the current version to work.

**Comment by neox387** [comment]
Not that much work but I have little time nowadays :D

Well if there is some Older version without The problem would be Nice to find out exactly what commit causes the problem

**Comment by Killerpretzel** [comment]
@H4R0 @neox387 Hey guys,

I ran some more tests. and I now think it's related to magnet links. After I disabled them in Sonarr, the problem disappeared.

After that I manually created some more downloads with magnet links and the problem crept back up almost immediatly. 

If you get the chance, you should check them out yourself.

**Comment by neox387** [comment]
try again with next PR should hopefully be ok now..

**Comment by Killerpretzel** [comment]
Hey @neox387 

thanks a lot, I really appreciate it!

After some tinkering with Docker, I gave the new version a spin.

Sadly, it's not working for me. 

I enabled magnet-links and encountered downloads without assigned categories shortly thereafter. I am still not 100% sure though, if it's really a problem with the magnet links, but I see very few. if any problems without them. 

I will disable the magnet-links again. It works quite well without them, but they are quite common nowadays. I suspect Sonarr has a lot less download-sources to work with, without them.

Take care.

**Comment by tuefekci** [comment]
I am using it with docker for about 6 months now works fine only sometimes it has probelms which i think are not related to docker restart in the application is enought to get it back on track.

thats my dockerfile
`FROM lsiobase/xenial.armhf
MAINTAINER Giacomo Tüfekci <kontakt@tuefekci.de>

# environment settings
ARG DEBIAN_FRONTEND="noninteractive"
ENV HOME="/config"

# Install
RUN apt-get update && apt-get -y install git python python-pip virtualenv 

# Copy this repo into place.
COPY . /srv/premiumizer

# virtualenv & pip
RUN cd /srv/premiumizer; virtualenv virtualenv; source virtualenv/bin/activate; pip install -r requirements.txt

# Expose apache.
EXPOSE 5000

# By default start up apache in the foreground, override with /bin/bash for interative.
CMD python /srv/premiumizer/premiumizer.py

VOLUME /srv/premiumizer /blackhole /downloads`

perhaps it helps somebody.

**Comment by kungfoolfighting** [comment]
It would be great to get an automatically built image for docker from the docker hub. As far as I can tell only the maintainer of the github account for premiumizer can link the docker hub automatic build.
I have spent quite some effort getting a good Dockerfile, however the whole process of keeping the Premiumizer container up-to-date with the github repository requires a lot of manual intervention this way.
My proposal is: I create a pull request for the dockerfile and the .dockerignore and the maintainer could then create an automatic build on docker hub.
That way users can use tools like 'Watchtower' to automatically update their containers when new commits have been made to the github repository.
Another benefit to the dockerfile being created in the repository and docker hub doing an automatic build is, that there doesn't have to be a git command in the Dockerfile (which is problematic) and also the users don't have to manually provide the repository as a build context to the docker build.

What do you think?

**Comment by piejanssens** [comment]
Sounds good to me!

**Comment by kungfoolfighting** [comment]
@piejanssens I had to wait till after work and do some small changes.
PR is up now. Hope you can use it.

**Comment by kungfoolfighting** [comment]
thanks  @piejanssens for that very fast work :+1: 
Do you know how to setup the automatic builds on hub.docker.com? And are you planning on doing that right now?

**Comment by kungfoolfighting** [comment]
docker run -d --name premiumizer --user 985:1009 -p 5000:5000 -v /etc/localtime:/etc/localtime:ro -v /etc/timezone:/etc/timezone:ro -v /opt/premiumizer_docker/config/settings.cfg:/premiumizer/settings.cfg -v /opt/premiumizer_docker/blackhole:/blackhole -v /opt/premiumizer_docker/download:/download premiumizer

**Comment by piejanssens** [comment]
It's already setup, but there appears to be a problem:
https://hub.docker.com/r/piejanssens/premiumizer/builds/bsrcwebigs3nzv3uj8ig5ro/

**Comment by piejanssens** [comment]
@kungfoolfighting I think the docker hub server doesn't support the 'chown' flag.

**Comment by kungfoolfighting** [comment]
humm.. that sucks..
Well i can change it for now, even though that will cost significant size in the image.
I will create a PR

**Comment by piejanssens** [comment]
<img width="867" alt="schermafbeelding 2017-12-21 om 21 19 20" src="https://user-images.githubusercontent.com/829098/34273460-af85be2a-e694-11e7-8bf2-6a678e905089.png">

Thanks @kungfoolfighting for the Dockerfile.

@bjsjr @disrupted @tuefekci @Killerpretzel Can you guys test? Anyone that could share his experience of this dockerfile on his synology?

**Comment by kungfoolfighting** [comment]
YAY! works like a charm here. Thanks a lot!

**Comment by Killerpretzel** [comment]
Hey @piejanssens @kungfoolfighting 

I gave the container a go. 

Unfortunately, I encountered a very simillar error message to the one in my own tests. Is this an Alpine based container? I never got Alpine to work and had to switch over to Ubuntu.

If this is any use to you, I am running Docker V.17.07.0-0367, on my Synology DS916+, with DSM 6.1.4-15217 Update 5.

The error message says:
"ValueError: Unable to determine the name of the local timezone -- you must explicitly specify the name of the local timezone. Please refrain from using timezones like EST to prevent problems with daylight saving time. Instead, use a locale based timezone name (such as Europe/Helsinki)."

[error.txt](https://github.com/piejanssens/premiumizer/files/1591312/error.txt)

![image](https://user-images.githubusercontent.com/33689789/34412614-92f60180-ebde-11e7-9016-49f1953fb335.png)

**Comment by piejanssens** [comment]
@Killerpretzel 
In the container add an 'environment variable' with the name 'TZ' and value of your time zone. See https://en.wikipedia.org/wiki/List_of_tz_database_time_zones column 'TZ'.

**Comment by kungfoolfighting** [comment]
@Killerpretzel 
I spent a lot of time getting the TZ stuff to work.
The trick lies within the run command.
Linuxes have very different ways of dealing with timezones. Alpine is a bit different to UBUNTU but in my opinion does it better.

In one of my previous comments I posted my run:

> docker run -d --name premiumizer --user 985:1009 -p 5000:5000 -v /etc/localtime:/etc/localtime:ro -v /etc/timezone:/etc/timezone:ro -v /opt/premiumizer_docker/config/settings.cfg:/premiumizer/settings.cfg -v /opt/premiumizer_docker/blackhole:/blackhole -v /opt/premiumizer_docker/download:/download premiumizer
You can see that I do two things: I link in the /etc/localtime file, which contains the actual data for the timezone, and then I also link in the /etc/timezone file.
This is according to https://wiki.alpinelinux.org/wiki/Setting_the_timezone.
Keep in mind that alpine linux is tricky with the"TZ" environment variable. I had to fight with that as well. The problem is that when you supply something like "Europe/Berlin" in the TZ variable, Alpine will not use /etc/localtime but rather /usr/share/zoninfo/Europe/Belin as its timezone source. This does not work, as I have not installed tzdata in the image, since it bloats the image and there are other ways of doing this. I also stumbled really hard over this.

**Comment by kungfoolfighting** [comment]
@piejanssens  This does not work in this image, I am afraid. See my previous comment on why.

**Comment by piejanssens** [comment]
@kungfoolfighting Thanks for clarifying. So if I understand your run command correctly @Killerpretzel could get it to work with the Docker UI in Synology's DSM by mapping `/etc/localtime` and `/etc/timezone`?

**Comment by kungfoolfighting** [comment]
@piejanssens 
Yeah that is correct. I don't know about the specifics of synology NASs though, so I have no idea if /etc/localtime and /etc/timezone exist the way they do in other Linuxes and if the Synology docker environment grants access to those. On my Arch Linux host for example, there is not /etc/timezone file, but the contents should only be the string "Europe/Berlin" in my case anyways, so I created a local file in my premiumizer_docker directory and linked that one to /etc/timezone.

**Comment by piejanssens** [comment]
Do you have an idea how we can make the image work with the `TZ` environment variable?

**Comment by kungfoolfighting** [comment]
@piejanssens 
yeah, you can install tzdata in the image.
It's up to you if you want to do it like that.
I personally didn't do it, because I always link in the hosts localtime anyways. That way all containers are always synced and I don't need to change stuff manually. 
Even if you want a timezone that is different to your hosts you can always so something like 
`-v /usr/share/zoneinfo/Europe/Berlin:/etc/localtime`
Using tzdata in the alpine image just adds a bunch of those zoneinfo files into the image and then you can use the TZ env var and alpine will use that var to point to the zoneinfo files.
The change to the docker file is minimal.

**Comment by piejanssens** [comment]
Would be good if we could make it work supporting both -v /etc/localtime and -e TZ.

**Comment by kungfoolfighting** [comment]
sure thing.  Will create a PR sometime soon.

**Comment by piejanssens** [comment]
@kungfoolfighting Awesome! It's nice to have someone on board who knows his Docker stuff! 😃

**Comment by kungfoolfighting** [comment]
Implemented the changes. Branch is here:
https://github.com/kungfoolfighting/premiumizer/commits/TZ
Waiting for https://github.com/piejanssens/premiumizer/issues/100 to be resolved before I create PR, though, so I can properly test.

**Comment by piejanssens** [comment]
@Killerpretzel Please try with the latest image. Thanks to @kungfoolfighting you can now use the TZ environment variable to set the timezone. 

Do let us know if it works out for you.

**Comment by Killerpretzel** [comment]
@piejanssens @kungfoolfighting 

Hey guys,

I got it to work, the environment variable did the trick. Thanks a lot!

However, after the Premiumizer-WebGUI started successfully, I tried to setup the access to my NAS-filesystem. This failed with a "permission denied" error, even though the container is set up in a simmilar way to my others, which work fine.

After some analysis, I found that the Docker image tries to access the files with the UID "6006". Unfortunatly, it doesn't seem to be possible to create users with specific UIDs within the Synology WebGUI and not even on the commandline. Additionally, the user management is set up in a way that makes it impossible to retain custom changes to this structure. That's what I gathered from some user comments, at least.

I had to allow the group "everyone" to access the files, this seems to work for now, but as you might imagine, it is a rather undesired approach.

With the Ubuntu image I used previously, this was no problem at all. What's the difference with Alpine? Also, I would have expected that Docker takes care of the permissions and that the behaviour remains the same, regardless of the used container.

**Comment by kungfoolfighting** [comment]
@Killerpretzel 
Ah yes the permissions.. They are always a bit complicated with Docker.
The thing is that it is generally recommended not to run services as root in a container.
That is why I added a premiumizer user with the UID of 6006 in the container.
The way that user management works in the container is the same as with your host, since you share the kernel. The kernel doesn't care where the UID comes from (host or container) and as such you share permissions between the host and container for the UIDs.
As you correctly stated, you could add a user with the UID 6006 to the host machine and that would work fine, but since Synology doesn't seem to allow you to do that, you have other possibilities: You can pass custom UID and GID in the run command (--user flag). Again, see my example RUN command that I posted earlier.
Of course these will then also be subject to access restrictions.
If you want to have the behavior that is exactly like you had it in your other container, you can pass the root uid to the run command and be done with it. But keep in mind that this is discouraged by the docker community.

Hope this helps.

**Comment by piejanssens** [comment]
In other Syno compatible docker containers I see that they use the environment variables PUID and GPID.
See https://drfrankenstein.co.uk/2017/03/28/setting-up-radarr-in-docker-on-a-synology-nas/ for an example. I assume we could do something similar @kungfoolfighting

**Comment by piejanssens** [comment]
Looking at https://github.com/linuxserver/docker-radarr there doesn't seem to be anything specific in the dockerfile to support the env variables setting PUID and GPID. This means that if @Killerpretzel would try setting them using the Syno Docker UI it should be OK?

**Comment by kungfoolfighting** [comment]
@piejanssens 
I had thought about doing this when I initially made the dockerfile, but decided against it.
The reason there is nothing in the dockerfile that you can see is that the infrastructure of the application running inside the container deals with it, not the dockerfile. radarr for example runs as a service inside the container, and the service uses the env vars to run the application as that user.
When I say "infrastructure", this is what I mean. The service files for example, or a run script for the application.
The reason why I decided against the env var approach was, that I would need to write a startup shell script to achieve that, and I wanted to keep it as simple as possible, seeing how the --user command seems to achieve the exact same thing.

Correct me, if you think there is a difference.

**Comment by kungfoolfighting** [comment]
BTW: I think a readme file for the dockerhub page is needed, like they have for most other projects. The amount of configuration that one can do for containers is always pretty big. And this thread shows that without the documentation, users will have issues.

**Comment by kungfoolfighting** [comment]
Oh and one more thing:
The reason why I prefer not to run as a service is that afaik the container will not exit as soon as your application exits or crashes, and it can therefore not automatically be restarted. So imho the best approach (if you REALLY must support the env vars) would be to have a startup script, which someone has to write :-)  )

**Comment by GrubbyDungles** [comment]
I just thought I'd add for anyone else struggling to get this working, setting the config as an environment variable rather than a volume got me up and running. I think this might be because I am using docker compose.

Using volume created a directory called `settings.cfg `and gave error,  

```
Traceback (most recent call last):
  File "premiumizer.py", line 59, in <module>
    active_interval = prem_config.getint('global', 'active_interval')
  File "/usr/local/lib/python2.7/ConfigParser.py", line 359, in getint
    return self._get(section, int, option)
  File "/usr/local/lib/python2.7/ConfigParser.py", line 356, in _get
    return conv(self.get(section, option))
  File "/usr/local/lib/python2.7/ConfigParser.py", line 330, in get
    raise NoSectionError(section)
ConfigParser.NoSectionError: No section: 'global'
```

**Comment by disrupted** [comment]
I need /config as a volume for my setup. also need to set UID & GID. Could the Dockerfile be updated to support that?

**Comment by misctempacc** [comment]
I have ARM based box with libreelec installed on it. I have set up functional Docker environment. How do I install Premiumizer Docker container? Will it work on ARM based machine? What will be Docker Run command?

**Comment by cron410** [comment]
@misctempacc You might have trouble running this on ARM. You will probably have to base this image off an ARM based image instead of Python:2.7 image as it is probably running x86 binaries. I am no expert in running docker on ARM.

**Comment by cremor** [comment]
@kungfoolfighting Thanks for your work on this, but is there any progess on (1) a readme that shows all required startup settings or (2) a solution for the UID problem for Synology users?

**Comment by kungfoolfighting** [comment]
>  is there any progess on (1) a readme that shows all required startup settings

I don't know. The maintainer could just add a readme file for the docker hub I think. Anyone can write that and create a pull request.

> (2) a solution for the UID problem for Synology users?

I don't know what problem you mean.

I posted how to get it working way above:

`docker run -d --name premiumizer --user 985:1009 -p 5000:5000 -v /etc/localtime:/etc/localtime:ro -v /etc/timezone:/etc/timezone:ro -v /opt/premiumizer_docker/config/settings.cfg:/premiumizer/settings.cfg -v /opt/premiumizer_docker/blackhole:/blackhole -v /opt/premiumizer_docker/download:/download premiumizer`

I don't know about any specific Synology issue. Most of the questions that were asked here had to do with people not using all the commands in that example run command I think, or not understanding what the --user directive means.
I don't know the specifics of Synology, so I don't know what you need to get it to work there.

If you have any more specific questions I am happy to answer them.

**Comment by GOvEy1nw** [comment]
Hey, I'm trying to set this up on my Synology NAS I've run into the following issue: 

```
11-17 15:28:03: INFO : Starting server on 0.0.0.0:5000
11-17 15:28:13: ERROR : premiumize.me login error: {"status":"error","message":"Not logged in. Please log in first."}
```

I had already set my port to something other than 5000. Any way to fix this?

**Comment by neox387** [comment]
it reads config from settings.cfg, so if that still says 5000 it's not changed, but guess this has something to do with docker ..

**Comment by GOvEy1nw** [comment]
> it reads config from settings.cfg, so if that still says 5000 it's not changed, but guess this has something to do with docker ..

Hey, thanks for the reply. How do I access that config file? I cant find it in my docker/premiumizer/config folder which is where I thought it would be.

**Comment by neox387** [comment]
Maybe this.. Comment above you said:

I had to manually download the settings.cfg.tpl and rename it to even start the container.

**Comment by Flobbes** [comment]
> The problem with Synology is that there is no way to specify the --user Docker parameter in the GUI. And there is also no way to create a user or group with a specific Id. Changing the Id of a user or group via Linux commands doesn't help either because the Synology system overwrites that again.
> As far as I know, there are only two options for Synology concerning access privileges: Run as root inside the container (not nice) or specify UID and GID with environment variables (which has to be supported by the container).

@cremor have you found a solution how to handle the "synology problem"?
I have exactly the same problem which breaks part of the automation of the premiumizer idea, because the postprocessing application like couchpotatoe can't process the file because of missing permissions.
So I have to manually execute several steps after the download.

I was thinking about running a cronjob on the folder setting the permissions to 777 for all downloaded files. But that is a really nasty workaround.

**Comment by cremor** [comment]
Right now I'm using my own Docker image. I've created a fork of this repository, changed the UID and GID in the code and let Docker Hub build a image based on that.
Of course that's far from ideal and I won't get any automatic updates that way, but at least it works.

**Comment by Flobbes** [comment]
I think I will stick to this solution as well, until maybe some day a general solution is found.
Is it this one https://hub.docker.com/r/cremor/premtest/ ?

Another approach to target this problem, could be to add an additional (optional) option to premiumizer to write files with 777 permissions. Don't know if the developers would approve such an additional option.

**Comment by cremor** [comment]
> I think I will stick to this solution as well, until maybe some day a general solution is found.
> Is it this one https://hub.docker.com/r/cremor/premtest/ ?

Yes, that's the one I'm using now. But it won't help you since it uses my UID/GID. I've only created it as a public repository because I couldn't find a way to add an image from a private repository to the Synology Docker GUI.

> What CPU Architecture does the synology nas use, x86/x64 or arm ?

I'm not 100% sure, but I think Synology only offers Docker on their x86/x64 models.

> All containers that deal with the media files have to use the same uid/gid (e.g premiumizer, sonarr, radarr, samba, minidlna etc.)

That's not true. I run a few containers that interact with each other (e.g. Sonarr and Premiumizer) and they work fine with different UIDs. They do share the same GID though, I've never tried it with different ones.

**Comment by Flobbes** [comment]
> I can upload my debian x86/x64 premiumizer image to dockerhub that deals with file permission, you could use it to test your case. I could also port it to arm.

I would give it a try, if you would upload it. x64 is fine.

> All containers that deal with the media files have to use the same uid/gid (e.g premiumizer, sonarr, radarr, samba, minidlna etc.)

I'm not using a container for the other apps, instead native synology packages. Then all have their own users and uids, but don't have any problems to interact. Although to be honest, I'm not sure how they handle the permissions (maybe they belong to the them group). Unfortunately the user management on the synology is a mess.

**Comment by cron410** [comment]
@cremor I use Backstroke to automatically make pull requests from a repo I don't control. This allowed me to work but Jeep receiving updates. Just need to approve and merge each one, that's just 2 mouse clicks any time there is an update to this repo.

**Comment by kungfoolfighting** [comment]
The correct way to do this is write a shell script that interprets the uid environment vars. But I am too lazy to write that because I don't have the environment where it would only work like that and also it makes the container more complicated just for this edge case. But feel free to add a shellscript and create a pullrequest for the changes to the docker file and the shell script.

**Comment by kungfoolfighting** [comment]
Seems like we have a working and feature rich docker container. lets close this issue?

**Comment by piejanssens** [comment]
I'm looking for some info on how to get it running on Synology DSM via the Docker UI. 
Which mount points are necessary? I get "container does not exist" when trying to start it.
Is there anyone that is running this on DSM?

**Comment by jan-auer** [comment]
@piejanssens I am - successfully so. 

**Environment Variables:**
Set `TZ` (e.g. `Europe/Berlin`), `PUID` and `PGID`, as well as `UMASK` (see below). 

**Mount Points:**
 - `/premiumizer/settings.cfg`: Some file that contains settings so that it survives container recreation
 - `/premiumizer/{torrents,downloads}`: A folder for downloads and the Torrent blackhole, as configured. Can be placed anywhere, obviously.
 - `/etc/profile.d/umask.sh`: A script to fix up the UMASK to make sure that downloads are also writable by other applications not running under the same user. This happens when using packages from SynoCommunity, which use their own user and group. Since this setup effectively requires 777 permissions:

```sh
# contents of mounted /etc/profile.d/umask.sh
umask 0000
```

The other ends of those mount points can be files and folders in any writable location, I chose a subfolder of `/volume1/video` (the default video folder).

**Caveats:**
I *think* when I set this up, it ended up setting permissions correctly for all subfolders that premiumizer might create in the download and blackhole folders. However, it's important to have the umask profile there **first**, or go in and chmod manually.

**Comment by cremor** [comment]
Yes, I'm running it on DSM. My settings:

```
   "env_variables" : [
[default ones not shown]
      {
         "key" : "PGID",
         "value" : "65537"
      },
      {
         "key" : "PUID",
         "value" : "1035"
      },
      {
         "key" : "TZ",
         "value" : "Europe/Berlin"
      }
   ],
   "network" : [
      {
         "driver" : "bridge",
         "name" : "bridge"
      }
   ],
   "network_mode" : "bridge",
   "port_bindings" : [
      {
         "container_port" : 5000,
         "host_port" : 5500,
         "type" : "tcp"
      }
   ],
   "volume_bindings" : [
      {
         "host_volume_file" : "/docker/Premiumizer/cfg/settings.cfg",
         "mount_point" : "/premiumizer/settings.cfg",
         "type" : "rw"
      },
      {
         "host_volume_file" : "/Downloads/Blackhole",
         "mount_point" : "/blackhole",
         "type" : "rw"
      }
   ],
```

**Comment by ujellyx** [comment]
Can someone, who got this working with blackhole, post screenshots of the volume-mappings inside Docker for this container?

**Comment by cremor** [comment]
Here is a screen that shows the same values that I've posted in the JSON settings format above:

![grafik](https://user-images.githubusercontent.com/116314/52649039-d3f6ca00-2ee7-11e9-86b5-562689bd5815.png)

**Comment by ujellyx** [comment]
> Here is a screen that shows the same values that I've posted in the JSON settings format above:
> 
> ![grafik](https://user-images.githubusercontent.com/116314/52649039-d3f6ca00-2ee7-11e9-86b5-562689bd5815.png)

Hm it seems then that mapping isn't the issue in my configuration. As PUID/PGID I need to manually set the administrator user and the administrator group respectively, right?

**Comment by jan-auer** [comment]
@ujellyx What is the problem that you are experiencing? Is premiumizer not able to read from the blackhole or are you not able to open the downloaded files?

**Comment by ujellyx** [comment]
> @ujellyx What is the problem that you are experiencing? Is premiumizer not able to read from the blackhole or are you not able to open the downloaded files?

The first one. It doesn't recognize new torrents added to the respective blackhole-folder. I can manually upload/drag&drop files into the frontend though, and the download works fine.

**Comment by jan-auer** [comment]
Assuming that you have configured the `/blackhole` folder correctly in your settings - can you open a shell in the container and see what user, group and access rights have in there? I'm assuming that premiumizer might not be able to open the files. 
Setting PGID and PUID to the values of your DSM user would likely help in this case.

**Comment by cremor** [comment]
> Hm it seems then that mapping isn't the issue in my configuration. As PUID/PGID I need to manually set the administrator user and the administrator group respectively, right?

I create a separate user for each Docker container, therefore the `PUID` has the id of that special user. And I also have a group that contains all that Docker users, so my `PGID` is set to that one.

**Comment by piejanssens** [comment]
We need instructions in our readme on a typical Docker installation.

**Comment by piejanssens** [comment]
@cremor Moving the settings file to a separate folder would simplify the process?

**Comment by cremor** [comment]
Disclaimer: Speculation follows, since I haven't tested it and don't use Docker that often to know this things for sure.

Wouldn't that lead to all binaries being in the mapped folder on the Docker host? I'd only want configuration files on my host system, all binaries should stay in the container.

Also, what about updates? If a new image (with new binaries) is pulled, would that even do anything? Wouldn't the container continue to use the old binaries from the volume?

**Comment by jan-auer** [comment]
While it's common practice to mount configuration directories, it is totally fine to just mount single files. You can see this in the screenshot @ujellyx posted or in my previous comments. Unless you're planning to introduce more configuration files, there is no disadvantage at the moment.

**Comment by cremor** [comment]
The problem with mounting single files lies with Synology's DSM. In their Docker GUI, you can only select existing files to be mounted. Therefore you have to manually copy the template settings file (and rename it), which would not be needed if we could mount a config-folder.

**Comment by ujellyx** [comment]
Agreed. Since you can't mount the "root"-folder of the container, it would help to have some kind of "config"-folder, just like many other docker images have (such as Sonarr, sabnzbd, etc.)

**Comment by piejanssens** [comment]
Project files have been restructured and there is now a separate 'conf' folder that you should mount.

**Comment by piejanssens** [comment]
Heads up. With next week's update we're switching our codebase and dependencies to Python 3.

**Comment by piejanssens** [comment]
Everything is configurable via the UI, therefore we shouldn't be a needing to mount the `/conf` directory or am I mistaken?

**Comment by cremor** [comment]
You might not *need* it, but it's always a good idea to map any configuration from Docker containers, otherwise it's lost when you reset or update a container.

**Comment by piejanssens** [comment]
@cremor Got it 👍 Anything we still need to improve to our docker config/support?

**Comment by neox387** [comment]
Ye like this in Docker file:
FROM python:3.7-alpine


There is also 3.9-alpine etc I saw..

**Comment by piejanssens** [comment]
Something is f*cked. Clean install of docker container on DSM (PUID and PGID set correctly).
Mapping:
- somefolder/premiumizer:/conf
- somefolder/downloads:/downloads
Start premiumizer and it runs, I can save config and it's persistent across reboots but there is nothing visible in `somefolder/premiumizer`. This means the file is conf file is written inside the docker but not at '/conf'.

**Comment by piejanssens** [comment]
OK guys I think it's way better now (see readme for new docker info)
Basically I did what I suggested above, but I also moved the logs and the .db to the mapped `/premiumizer/conf` location.

**Next up Dockerfile** I need your help.
We should simplify and improve the dockerfile.
We should try and get rid of the chmod 777.
We should try to get rid of the `premiumizer` user.
Do we move to alpine 3.9?
Now that TZ works, can we not get rid of this line?
`COPY --from=0 /usr/share/zoneinfo /usr/share/zoneinfo/`

**Comment by cremor** [comment]
I've justed tested the Docker image that was build for commit ad2b8ced3d1ed8c4eb11066e2370e8eb3b64e09f on a Synology NAS. Good work with the new configuration options and updated readme, I had no problems with it!

**Comment by onemico** [comment]
Hi,  this might be a silly question, but had this working fine on my OMV instance when it was an old laptop, but I just moved it all to a Raspberry Pi 4 and now the docker container won't start with the error 

`standard_init_linux.go:211: exec user process caused "exec format error"`

Now after Googling I think it is because the container doesn't support arm processors? Is anyone able to confirm this, and if so, is their an alternative way to get it working within a docker container on arm?

**Comment by neox387** [comment]
@onemico #208 he seems to run on PI so should work i guess =D

**Comment by onemico** [comment]
> @onemico #208 he seems to run on PI so should work i guess =D

Hmmm i cannot get the Docker to run at all, have ended up just running it in a venv for now.

**Comment by tuefekci** [comment]
> Hi, this might be a silly question, but had this working fine on my OMV instance when it was an old laptop, but I just moved it all to a Raspberry Pi 4 and now the docker container won't start with the error
> 
> `standard_init_linux.go:211: exec user process caused "exec format error"`
> 
> Now after Googling I think it is because the container doesn't support arm processors? Is anyone able to confirm this, and if so, is their an alternative way to get it working within a docker container on arm?

https://github.com/piejanssens/premiumizer/issues/255#issuecomment-636320196
